google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java

```diff
@@ -1239,11 +1259,19 @@ void backoffSleep(Context context, long backoffMillis) {
     @Override
     public <T> T run(TransactionCallable<T> callable) {
       try (Scope s = tracer.withSpan(span)) {
+        if (blockNestedTxn) {
+          hasPendingTransaction.set(Boolean.TRUE);
+        }
+
         return runInternal(callable);
       } catch (RuntimeException e) {
         TraceUtil.endSpanWithFailure(span, e);
         throw e;
       } finally {
+        if (blockNestedTxn) {
+          hasPendingTransaction.set(Boolean.FALSE);
```

Comment:
I'd use `hasPendingTransaction.remove()` instead. OpenJDK implementation (and possibly others) use `Map<Thread, Value>` underneath, `remove` will actually remove the mapping, setting this to false is a bit of a memory leak.

Is there a chance someone might change `allowNestedTransaction` while a transaction is running? If there is, We might want to remove unconditionally.

Reply:
Thanks for the info on the memory leak. 
Although unlikely, it _is_ possible in theory to change the flag so best to handle it I guess.

Good catch. Will make both the changes.